scipy/optimize/_trustregion_constr/canonical_constraint.py

```diff
@@ -133,7 +133,7 @@ def hess(x, v_eq, v_ineq):
                 index_ineq += c.n_ineq
 
             def matvec(p):
-                result = np.zeros_like(p)
+                result = np.zeros_like(p, dtype=float)
```

Comment:
What if the input is `float32`?

Reply:
In the test, even with `p.dtype=int8`, the `result` is already `float64` giving the error in `+=`

Reply:
Which test? How is it even getting an `int8` array to begin with? Since the `dtype` is explicitly set in the constructor a few lines down, the dtype-guessing code shouldn't activate.

But since it *is* explicitly instantiated with `dtype=float`, this change is a good one. I'm just puzzled as to how you encountered it.

Reply:
the parameter p is an input to the LinearOperator and could in theory be of any type. My change made it int8 and resulted in a type mismatch error adding int8 and float64 with += failing a test. I have not paid attention to anything else around. You could find the error in the log.

Reply:
Ah, it's not this `LinearOperator` that's failing to determine the `dtype`, but the one inside of `LagrangianHessian` which is composed from these.

I'm satisfied that this is a good change now that I know where the failure arose.